### PR TITLE
Multi-arch images build via drivah and serverless-docker-builder

### PR DESCRIPTION
**Pull request by thbkrkr**
This change revisits the build of ECK's container images by making them drivah-able and use the `serverless-docker-builder`.

The main thing is:
```yaml
  - group: build operator
    key: "operator-image-build"
    steps:
      - label: ":buildkite:"
        agents:
          image: "docker.elastic.co/ci-agent-images/serverless-docker-builder:0.0.5"
        commands: make -C /agent generate-docker-images
        env:
          DOCKER_REGISTRY_VAULT_PATH: secret/ci/elastic-cloud-on-k8s/docker-registry-elastic
          PRE_BUILD_SCRIPT: ".buildkite/scripts/build/pre-build-operator.sh"
          DRIVAH_BUILD_PATH: "build"
          RECURSIVE: true

  - group: build e2e-tests
    key: "e2e-tests-image-build"
    steps:
      - label: ":buildkite:"
        agents:
          image: "docker.elastic.co/ci-agent-images/serverless-docker-builder:0.0.5"
        commands: make -C /agent generate-docker-images
        env:
          DOCKER_REGISTRY_VAULT_PATH: secret/ci/elastic-cloud-on-k8s/docker-registry-elastic
          DRIVAH_BUILD_PATH: "test/e2e"
```

The `serverless-docker-builder` agent step generates 3 steps to build the arm and the amd images and push the manifest. It executes `drivah build $DRIVAH_BUILD_PATH --push` which discovers all `Dockerfile` and `drivah.toml*` in the given path and build the associated container images using the buildah engine, in a `k8s` agent for amd and a `vm` agent for arm.
This means that we don't have `make` or `go` when `drivah build`.

In order to be able to use drivah to build eck, we now fetch the license public key first and then we `go-generate` in Docker during the build.

I made that we can still invoke the build of images from the `Makefile` to dev purposes.

```sh
# to install drivah (requires python 3.9+, git, buildah or docker)
pip install git+ssh://git@github.com/elastic/drivah@0.24.0
```

### e2e-tests image

In CI, `drivah.toml.sh` is directly called by `drivah build`.

In dev, you can:
```sh
# build without license public key
GO_TAGS= make drivah-build-e2e

# build with license public key (default) (this uses your current arch (arm if m1) and the docker engine)
make drivah-build-e2e 

# build/push amd/arm from an arm? machine using docker buildx
make docker-push-e2e
```

### operator image(s)

A new `BUILD_FLAVORS` variables is added to select the images to build (`eck,eck-dev,eck-fips,eck-ubi8,eck-ubi8-fips`). 

It is used by `build/gen-drivah.toml.sh` that adds an intermediate step compared to the e2e tests to generate the drivah.toml config.

```sh
> tree build
build
├── Dockerfile
├── Dockerfile-ubi
└── gen-drivah.toml.sh

> BUILD_FLAVORS=eck,eck-dev,eck-fips,eck-ubi8,eck-ubi8-fips make drivah-generate
# -- gen-drivah-config BUILD_FLAVORS=eck,eck-dev,eck-fips,eck-ubi8,eck-ubi8-fips
# -- build: docker.elastic.co/eck-dev/eck-operator-krkr:2.9.0-SNAPSHOT-ad0a893b
# -- build: docker.elastic.co/eck-dev/eck-operator-krkr:2.9.0-SNAPSHOT-ad0a893b-dev
# -- build: docker.elastic.co/eck-dev/eck-operator-krkr-fips:2.9.0-SNAPSHOT-ad0a893b
# -- build: docker.elastic.co/eck-dev/eck-operator-krkr-ubi8:2.9.0-SNAPSHOT-ad0a893b
# -- build: docker.elastic.co/eck-dev/eck-operator-krkr-ubi8-fips:2.9.0-SNAPSHOT-ad0a893b

> tree build
build
├── Dockerfile
├── Dockerfile-ubi
├── dev-license.key
├── eck
│   ├── Dockerfile
│   └── drivah.toml
├── eck-dev
│   ├── Dockerfile
│   └── drivah.toml
├── eck-fips
│   ├── Dockerfile
│   └── drivah.toml
├── eck-ubi8
│   ├── Dockerfile
│   └── drivah.toml
├── eck-ubi8-fips
│   ├── Dockerfile
│   └── drivah.toml
├── gen-drivah.toml.sh
└── license.key

6 directories, 15 files
```

Then drivah will build all images for you:
```sh
> drivah build build
Using Docker container engine.
Scanning /Users/krkr/dev/src/github.com/elastic/cloud-on-k8s/build...
Found 5 valid image definitions.
...
```

In CI, `.buildkite/scripts/build/pre-build-operator.sh` is executed first by the builder agent. It is a wrapper that will run `build/gen-drivah.toml.sh` after setting image name, image tag and build flavors variables depending on the Buildkite trigger.

`.buildkite/scripts/build/pre-build-operator.sh` is still usable in dev to for example test the generation for a given trigger. Note also that when preparing the build flavors, we do a composition of `-dev`, `-fips` and `-ubi8`, so the order is not important and you can do:

```sh
> TRIGGER=merge-main BUILD_FLAVORS=eck-dev-ubi8-fips .buildkite/scripts/build/pre-build-operator.sh
# -- pre-build-operator TRIGGER=merge-main
# -- gen-drivah-config BUILD_FLAVORS=eck-dev-ubi8-fips
# -- build: docker.elastic.co/eck-snapshots/eck-operator-ubi8-fips:2.9.0-SNAPSHOT-ad0a893b-dev
```

In dev, you can directly run `build/gen-drivah.toml.sh` before `drivah build` to generate the `drivah.toml`:
```sh
# only generate the files and directories in build/
> make drivah-generate-operator

# generate and build
> make drivah-build-operator

# build/push amd/arm from an arm? machine using docker buildx
make docker-push-operator
```

We could have build e2e tests and operator images with one invocation of the serverless-docker-builder but we would lose ~2-3min. I preferred keep the two builds in //.

I've removed the dependency when running the e2e tests to not wait for the end of the builds for speed. If this becomes problematic, we can revert this behaviour.

With this new build, the total PR pipeline time is reduced by around 25%, from around 20 minutes to around 15 minutes.

### CODE REVIEW

**Comment by thbkrkr** [comment]
buildkite test this

File: Makefile [hunk 1]
```diff
@@ -362,93 +351,34 @@ switch-kind:
...
-		--build-arg VERSION='$(VERSION)' \
-		--platform $(BUILD_PLATFORM) \
-		-t $(OPERATOR_IMAGE) .
-	buildah push \
-		--storage-driver=vfs \
-		$(OPERATOR_IMAGE)
+# standard way to build operator image(s) using drivah
+export BUILD_FLAVORS ?= dev
+drivah-build-operator: drivah-generate
+	drivah build ./build
```
Comment (pebrc) [inline Makefile:377]: Something I ran into when building locally through `drivah`

```
er:  > [builder 8/9] COPY build/license.key /license.key:
⚠️  Docker: ------
⚠️  Docker: tmph_dbdkyw:24
⚠️  Docker: --------------------
⚠️  Docker:   22 |     COPY pkg/ pkg/
⚠️  Docker:   23 |     COPY cmd/ cmd/
⚠️  Docker:   24 | >>> COPY $LICENSE_PUBKEY_PATH /license.key
⚠️  Docker:   25 |     
⚠️  Docker:   26 |     # Build
⚠️  Docker: --------------------
⚠️  Docker: ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref 300e217c-0ac1-4f9f-b498-54e1f5adee6b::b118lvsqa95wl97f8fqsagh5m: "/build/license.key": not found
```

Somehow the build context does not contain the `license.key`, when I build manually with docker from the generated Dockerfile it seems to work though. The file is present in `build/license.key`

**Review by pebrc — COMMENTED** [review]
LGTM, I ran into one snag when building locally. I really like the clean structure!

File: test/e2e/Dockerfile [hunk 2]
```diff
@@ -1,31 +1,34 @@
...
 COPY config/ config/
 COPY test/ test/
 
+# non-root user to support restricted security policies
+RUN chown -R 1001 /go/src/github.com/elastic/cloud-on-k8s
+USER 1001
+
+# go generate pub key
+COPY test/e2e/license.key /license.key
+RUN LICENSE_PUBKEY=/license.key go generate -tags="${GO_TAGS}" ./pkg/... ./cmd/...
```
[the remark this answers is not in the document]
Comment (thbkrkr) [reply test/e2e/Dockerfile:32]: When we `made go-generate` dependency of `make publish-e2e-multiarch-image`:

https://github.com/elastic/cloud-on-k8s/blob/1201379ee186abb13b437c147b97cad0abac1165/.buildkite/pipeline-build.yml#L10

https://github.com/elastic/cloud-on-k8s/blob/1201379ee186abb13b437c147b97cad0abac1165/cmd/license-initializer/main.go#L37

```
# in a gcp vm
# in a container docker.elastic.co/ci-agent-images/cloud-k8s-operator/buildkite-agent:d8e4780b
# where go, make, docker buildx are installed

controller-gen webhook/crd && kustomize &&  /hack/manifest-gen/manifest-gen.sh # (generate-manifests)
go generate -tags='$(GO_TAGS)' ./pkg/... ./cmd/... # (go-generate) 
└ go run ../../../../cmd/license-initializer/main.go --out zz_generated.pubkey.go <<<<<< HERE
hack/config-extractor/extract.sh # (generate-config-file)
hack/docker.sh -l -m $(OPERATOR_IMAGE) (docker-multiarch-build)
docker buildx build . \ 
	--progress=plain \
	--build-arg GO_LDFLAGS='$(GO_LDFLAGS)' \
	--build-arg GO_TAGS='$(GO_TAGS)' \
	--build-arg VERSION='$(VERSION)' \
	--platform $(BUILD_PLATFORM) \
	-t $(OPERATOR_IMAGE) \
	--push
```

File: Makefile (hunk 1 above)
Comment (thbkrkr) [reply Makefile:377]: `build/license.key` is created by `gen-drivah.toml.sh`. You could try to `set -x` in it to see what happens.

If `BUILD_FLAVORS=dev`, we create a fake license, which will not be used but the Dockerfile wants to copy a file.
https://github.com/elastic/cloud-on-k8s/blob/067621980a8af4cb860b6929316fa275fa828842/build/gen-drivah.toml.sh#L75
else it's read from Vault:
https://github.com/elastic/cloud-on-k8s/blob/067621980a8af4cb860b6929316fa275fa828842/build/gen-drivah.toml.sh#L94-L99

If you have a Vault error, you should see an error : `Error reading secret/..: Error making API request`.

Comment (pebrc) [reply Makefile:377]: The fake license file has been generated fine. But for some reason it cannot be copied in the when building the image.

File: .buildkite/scripts/common/trigger.sh [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            return 0
+        fi
+    done
+
+    echo unknown
+    return 1
+}
+
+is_tag-final() {
+    [[ "${BUILDKITE_TAG:-}" =~ ^v?[0-9]+\.[0-9]+\.[0-9]+$ ]] && return 0 || return 1
```
[the remark this answers is not in the document]
Comment (thbkrkr) [reply .buildkite/scripts/common/trigger.sh]: You're right!

**Review by pebrc — APPROVED** [review]
LGTM (problem was a mix of drivah versions on my machine)

**Comment by kezhenxu94** [comment]
> # to install drivah (requires python 3.9+, git, buildah or docker)
> pip install git+ssh://git@github.com/elastic/drivah@0.24.0

Hi @thbkrkr , is `elastic/drivah` close sourced? How can we build the multi-arch docker images on our own if `elastic/drivah` is not available to us?

**Comment by thbkrkr** [comment]
> Hi @thbkrkr , is `elastic/drivah` internal tool in elastic? How can we build the multi-arch docker images on our own if `elastic/drivah` is not available to us?

Yes, `elastic/drivah` is private, it is a tool that provides full automation for building Docker/OCI container images in bulk. It uses `buildah` or `docker` to build images.

You can still build multi-arch operator images using `docker buildx build` and its [Dockerfile](https://github.com/elastic/cloud-on-k8s/blob/main/build/Dockerfile). There is a `docker-push-operator` target in the `Makefile` that does this. You may want to add more platform in `BUILD_PLATFORM`.

https://github.com/elastic/cloud-on-k8s/blob/0d18fcb14913c0ab749edf53a423ac6c2408340d/Makefile#L362-L371

**Pull request merged**